cmd/soroban-cli/src/signer/types.rs

```diff
@@ -0,0 +1,183 @@
+use crossterm::event::{read, Event, KeyCode};
+use ed25519_dalek::ed25519::signature::Signer;
+use sha2::{Digest, Sha256};
+
+use crate::{
+    config::network::Network,
+    xdr::{
+        self, DecoratedSignature, Limits, Signature, SignatureHint, TransactionEnvelope,
+        TransactionSignaturePayload, TransactionSignaturePayloadTaggedTransaction,
+        TransactionV1Envelope, WriteXdr,
+    },
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Contract addresses are not supported to sign auth entries {address}")]
+    ContractAddressAreNotSupported { address: String },
+    #[error(transparent)]
+    Ed25519(#[from] ed25519_dalek::SignatureError),
+    #[error("Missing signing key for account {address}")]
+    MissingSignerForAddress { address: String },
+    #[error(transparent)]
+    Xdr(#[from] xdr::Error),
+    #[error(transparent)]
+    Rpc(#[from] crate::rpc::Error),
+    #[error("User cancelled signing, perhaps need to remove --check")]
+    UserCancelledSigning,
+    #[error("Only Transaction envelope V1 type is supported")]
+    UnsupportedTransactionEnvelopeType,
+}
+
+/// Calculate the hash of a Transaction
+pub fn transaction_hash(
+    txn: &xdr::Transaction,
+    network_passphrase: &str,
+) -> Result<[u8; 32], Error> {
+    let signature_payload = TransactionSignaturePayload {
+        network_id: hash(network_passphrase),
+        tagged_transaction: TransactionSignaturePayloadTaggedTransaction::Tx(txn.clone()),
+    };
+    let hash = Sha256::digest(signature_payload.to_xdr(Limits::none())?).into();
+    Ok(hash)
+}
+
+/// A trait for signing arbitrary byte arrays
+#[async_trait::async_trait]
+pub trait Blob {
+    /// Sign an abritatry byte array
+    async fn sign_blob(&self, blob: &[u8]) -> Result<Vec<u8>, Error>;
+}
+
+#[async_trait::async_trait]
+pub trait TransactionHash {
+    /// Sign a transaction hash with the given source account
+    /// # Errors
+    /// Returns an error if the source account is not found
+    async fn sign_txn_hash(
+        &self,
+        source_account: &stellar_strkey::ed25519::PublicKey,
+        txn: [u8; 32],
+    ) -> Result<DecoratedSignature, Error>;
+}
+#[async_trait::async_trait]
+impl<T> TransactionHash for T
+where
+    T: Blob + Send + Sync,
+{
+    async fn sign_txn_hash(
+        &self,
+        source_account: &stellar_strkey::ed25519::PublicKey,
+        txn: [u8; 32],
+    ) -> Result<DecoratedSignature, Error> {
+        eprintln!(
+            "{} about to sign hash: {}",
+            source_account.to_string(),
+            hex::encode(txn)
+        );
```

Comment:
This log line feels out of place now that we've moved to using the Print object. There'd need to be someway to pass the print object down to print here, or better move this print to a higher place. Printing here is probably not the right layer.

Reply:
This has also been moved to the top level signer and not in the default implementation. In fact there is no longer a default implementation rather a blanket impl.